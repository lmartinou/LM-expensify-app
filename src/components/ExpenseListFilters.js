import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };
    onDatesChange = ( { startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onSortChange = (e) => {
        if (e.target.value === 'date') this.props.sortByDate();
        else this.props.sortByAmount();
    };
    onFocusChange = (calendarFocused) => this.setState({ calendarFocused });
    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                    startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                    endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                    endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                    //additional
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                    displayFormat={"DD-MMM-YYYY"}
                    showClearDates={true}
                />
            </div>
        );
    }
}

const mapStoreToProps = (state) => (
    {
        filters: state.filters
    }
);

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStoreToProps, mapDispatchToProps)(ExpenseListFilters);
