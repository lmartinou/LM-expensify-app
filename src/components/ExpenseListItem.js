import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';


const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <div>
    <h3>
        <Link to={`/edit/${id}`}>
            {description}
        </Link>
        
    </h3>
    <p>{amount} - {moment(createdAt).fromNow()}, on {moment(createdAt).format('DD-MMM-YYYY')}</p>        
    </div>
);

export default ExpenseListItem;