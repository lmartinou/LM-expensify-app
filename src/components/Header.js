import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>    
        <h1>
            Expensify
    </h1>
        <p>
            <NavLink to="/" activeClassName="is-active" exact={true}>|Home| </NavLink>
            <NavLink to="/create" activeClassName="is-active"> |Add Expense| </NavLink>
            <NavLink to="/edit" activeClassName="is-active"> |Edit Expense| </NavLink>
            <NavLink to="/help" activeClassName="is-active"> |Help</NavLink>
        </p>    
    </header>
);

export default Header;

//removed:
//<NavLink to="/test" activeClassName="is-active"> |Dummy</NavLink>