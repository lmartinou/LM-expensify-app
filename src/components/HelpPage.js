import React from 'react';

const HelpPage = () => (
    <div>
        This is the HelpPage page from the router
    </div>
);

export default HelpPage;