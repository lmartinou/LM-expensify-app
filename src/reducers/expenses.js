//Expenses reducer
const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id !== action.id); //instead of (expense) =>, i use destructuring of the expense object, taking out the id
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        default:
            return state;
    };
};

export default expensesReducer;