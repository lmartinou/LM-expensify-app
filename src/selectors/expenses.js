import moment from 'moment';
//get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => { //spread the filters object
    return expenses.filter( (expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;        
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = (!text) || expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;        
    }).sort( (a, b) => {
        if (sortBy === 'date') { //return latest first            
            if (a.createdAt < b.createdAt) return 1;
            else if (a.createdAt > b.createdAt) return -1;
            else return 0;
        } else { //bring more expensive first
            if (a.amount < b.amount) return 1;
            else if (a.amount > b.amount) return -1;
            else return 0;
        }
    });
};