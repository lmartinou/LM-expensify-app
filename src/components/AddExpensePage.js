import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
};


const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    };
};

//the first argument in connect is always the mapStateToProps function.
//the second is the mapDispatchToProps. this is mapping of dispatch actions to enable shorter code and testability
export default connect(undefined, mapDispatchToProps)(AddExpensePage);