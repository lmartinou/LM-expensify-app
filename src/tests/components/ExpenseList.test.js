import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('should render the ExpenseList component', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render the ExpenseList with empty array', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});