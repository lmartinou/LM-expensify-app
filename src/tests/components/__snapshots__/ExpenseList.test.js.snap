// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the ExpenseList component 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note="my rent"
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit card"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`should render the ExpenseList with empty array 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;
